File: hermes-plugin/lore_memory/formatters.py
"""
Formatters - Output formatting for Lore memory data
"""

from typing import Any, List, Dict, Optional


def format_node(data: Dict) -> str:
    """Format a memory node for display"""
    node = data.get("node", {})
    children = data.get("children", [])
    
    lines = []
    lines.append(f"URI: {node.get('uri', '')}")
    
    if node.get("node_uuid"):
        lines.append(f"Node UUID: {node['node_uuid']}")
    
    lines.append(f"Priority: {node.get('priority', '')}")
    
    if node.get("disclosure"):
        lines.append(f"Disclosure: {node['disclosure']}")
    
    if node.get("aliases"):
        lines.append(f"Aliases: {', '.join(node['aliases'])}")
    
    lines.append("")
    lines.append(node.get("content", "(empty)"))
    
    if children:
        lines.append("")
        lines.append("Children:")
        for child in children:
            lines.append(f"- {child.get('uri', '')} (priority: {child.get('priority', '')})")
            if child.get("content_snippet"):
                lines.append(f"  {child['content_snippet']}")
    
    if node.get("glossary_keywords"):
        lines.append("")
        lines.append(f"Glossary keywords: {', '.join(node['glossary_keywords'])}")
    
    return "\n".join(lines)


def format_boot_view(data: Dict) -> str:
    """Format boot memory view"""
    core_memories = data.get("core_memories", [])
    recent_memories = data.get("recent_memories", [])
    failed = data.get("failed", [])
    loaded = data.get("loaded", len(core_memories))
    total = data.get("total", len(core_memories))

    lines = []
    lines.append("# Core Memories")
    lines.append(f"# Loaded: {loaded}/{total} memories")
    lines.append("")

    if failed:
        lines.append("## Failed to load:")
        lines.extend(failed)
        lines.append("")

    if core_memories:
        client_boot_memories = [memory for memory in core_memories if memory.get("scope") == "client"]
        lines.append("## Fixed boot baseline:")
        lines.append("")
        lines.append("Lore boot deterministically loads three global startup nodes inside Lore:")
        lines.append("- core://agent — workflow constraints")
        lines.append("- core://soul — style / persona / self-definition")
        lines.append("- preferences://user — stable user definition / durable user context")
        lines.append("")
        if client_boot_memories:
            lines.append(
                "This boot view also includes the active client-specific agent node:"
                if len(client_boot_memories) == 1
                else "This boot view also includes the client-specific agent nodes:"
            )
            for memory in client_boot_memories:
                lines.append(f"- {memory.get('uri', '')} — {memory.get('boot_role_label', 'client-specific agent constraints')}")
            lines.append("")

        for memory in core_memories:
            lines.append(f"### {memory.get('uri', '')}")
            if memory.get("boot_role_label"):
                lines.append(f"Role: {memory['boot_role_label']}")
            if memory.get("boot_purpose"):
                lines.append(f"Purpose: {memory['boot_purpose']}")
            if memory.get("priority") is not None:
                lines.append(f"Priority: {memory['priority']}")
            if memory.get("disclosure"):
                lines.append(f"Disclosure: {memory['disclosure']}")
            if memory.get("node_uuid"):
                lines.append(f"Node UUID: {memory['node_uuid']}")
            lines.append("")
            lines.append(memory.get("content", "(empty)"))
            lines.append("")
    else:
        lines.append("(No core memories loaded.)")

    if recent_memories:
        lines.append("---")
        lines.append("")
        lines.append("# Recent Memories")
        for memory in recent_memories:
            meta = []
            if memory.get("priority") is not None:
                meta.append(f"priority: {memory['priority']}")
            if memory.get("created_at"):
                meta.append(f"created: {memory['created_at']}")
            suffix = f" ({', '.join(meta)})" if meta else ""
            lines.append(f"- {memory.get('uri', '')}{suffix}")
            if memory.get("disclosure"):
                lines.append(f"  Disclosure: {memory['disclosure']}")

    return "\n".join(lines).strip()


def format_search_results(results: List[Dict], meta: Optional[Dict] = None) -> str:
    """Format search results"""
    if not results:
        return "No matching memories found."
    
    lines = []
    for idx, item in enumerate(results, 1):
        parts = [f"{idx}. {item.get('uri', '')} (priority: {item.get('priority', '')}"]
        
        if item.get("score") is not None:
            parts.append(f"score: {item['score']:.3f}")
        
        if item.get("matched_on"):
            parts.append(f"via: {'+'.join(item['matched_on'])}")
        
        lines.append(f"{', '.join(parts)})")
        lines.append(f"   {item.get('snippet', '')}")
    
    text = "\n".join(lines)
    
    if meta and meta.get("semantic_error"):
        text += f"\n\nSemantic fallback skipped: {meta['semantic_error']}"
    
    return text


def format_recall_block(items: List[Dict], session_id: Optional[str] = None, query_id: Optional[str] = None) -> str:
    """Format recall results as a block"""
    if not items:
        return ""
    
    attrs = []
    if session_id:
        attrs.append(f'session_id="{session_id}"')
    if query_id:
        attrs.append(f'query_id="{query_id}"')
    
    lines = [f"<recall {' '.join(attrs)}>" if attrs else "<recall>"]
    
    for item in items:
        score = item.get("score_display")
        if score is not None:
            score_str = f"{score:.2f}"
        else:
            score_str = str(item.get("score", ""))
        
        cues = item.get("cues", [])
        cue_parts = []
        cue_parts.extend(str(c).strip() for c in cues[:3] if c)
        cue_text = " · ".join(cue_parts)
        
        line = f"{score_str} | {item.get('uri', '')}"
        if cue_text:
            line += f" | {cue_text}"
        lines.append(line)
    
    lines.append("</recall>")
    return "\n".join(lines)


def format_domains(domains: List[Dict]) -> str:
    """Format domain list"""
    if not domains:
        return "No domains found."
    
    lines = [
        f"- {d.get('domain', '')} ({d.get('root_count', 0)}) — open root with lore_get_node uri=\"{d.get('domain', '')}://\" nav_only=true"
        for d in domains
    ]
    return "\n".join(lines)

File: pi-extension/formatters.ts
const DEFAULT_RECALL_SCORE_PRECISION = 2;

export function formatNode(data: any) {
  const node = data?.node || {};
  const children = Array.isArray(data?.children) ? data.children : [];
  const lines: string[] = [];
  lines.push(`URI: ${node.uri || ''}`);
  if (node.node_uuid) lines.push(`Node UUID: ${node.node_uuid}`);
  lines.push(`Priority: ${node.priority ?? ''}`);
  if (node.disclosure) lines.push(`Disclosure: ${node.disclosure}`);
  if (Array.isArray(node.aliases) && node.aliases.length > 0) {
    lines.push(`Aliases: ${node.aliases.join(', ')}`);
  }
  lines.push('');
  lines.push(node.content || '(empty)');
  if (children.length > 0) {
    lines.push('');
    lines.push('Children:');
    for (const child of children) {
      lines.push(`- ${child.uri} (priority: ${child.priority ?? ''})`);
      if (child.content_snippet) lines.push(`  ${child.content_snippet}`);
    }
  }
  if (Array.isArray(node.glossary_keywords) && node.glossary_keywords.length > 0) {
    lines.push('');
    lines.push(`Glossary keywords: ${node.glossary_keywords.join(', ')}`);
  }
  return lines.join('\n');
}

export function formatBootView(data: any) {
  const coreMemories = Array.isArray(data?.core_memories) ? data.core_memories : [];
  const recentMemories = Array.isArray(data?.recent_memories) ? data.recent_memories : [];
  const failed = Array.isArray(data?.failed) ? data.failed : [];
  const loaded = Number.isFinite(data?.loaded) ? data.loaded : coreMemories.length;
  const total = Number.isFinite(data?.total) ? data.total : coreMemories.length;
  const clientBootMemories = coreMemories.filter((memory: any) => memory?.scope === 'client');
  const lines: string[] = [];

  lines.push('# Core Memories');
  lines.push(`# Loaded: ${loaded}/${total} memories`);
  lines.push('');

  if (failed.length > 0) {
    lines.push('## Failed to load:');
    lines.push(...failed);
    lines.push('');
  }

  if (coreMemories.length > 0) {
    lines.push('## Fixed boot baseline:');
    lines.push('');
    lines.push('Lore boot deterministically loads three global startup nodes inside Lore:');
    lines.push('- core://agent — workflow constraints');
    lines.push('- core://soul — style / persona / self-definition');
    lines.push('- preferences://user — stable user definition / durable user context');
    lines.push('');
    if (clientBootMemories.length > 0) {
      lines.push(clientBootMemories.length === 1
        ? 'This boot view also includes the active client-specific agent node:'
        : 'This boot view also includes the client-specific agent nodes:');
      for (const memory of clientBootMemories) {
        lines.push(`- ${memory?.uri || ''} — ${memory?.boot_role_label || 'client-specific agent constraints'}`);
      }
      lines.push('');
    }
    for (const memory of coreMemories) {
      lines.push(`### ${memory?.uri || ''}`);
      if (memory?.boot_role_label) lines.push(`Role: ${memory.boot_role_label}`);
      if (memory?.boot_purpose) lines.push(`Purpose: ${memory.boot_purpose}`);
      if (Number.isFinite(memory?.priority)) lines.push(`Priority: ${memory.priority}`);
      if (memory?.disclosure) lines.push(`Disclosure: ${memory.disclosure}`);
      if (memory?.node_uuid) lines.push(`Node UUID: ${memory.node_uuid}`);
      lines.push('');
      lines.push(memory?.content || '(empty)');
      lines.push('');
    }
  } else {
    lines.push('(No core memories loaded.)');
  }

  if (recentMemories.length > 0) {
    lines.push('---');
    lines.push('');
    lines.push('# Recent Memories');
    for (const memory of recentMemories) {
      const meta: string[] = [];
      if (Number.isFinite(memory?.priority)) meta.push(`priority: ${memory.priority}`);
      if (memory?.created_at) meta.push(`created: ${memory.created_at}`);
      const suffix = meta.length > 0 ? ` (${meta.join(', ')})` : '';
      lines.push(`- ${memory?.uri || ''}${suffix}`);
      if (memory?.disclosure) lines.push(`  Disclosure: ${memory.disclosure}`);
    }
  }

  return lines.join('\n').trim();
}

export function readCueList(item: any) {
  const cues = Array.isArray(item?.cues) ? item.cues : [];
  const cleaned = cues.map((x: any) => String(x || '').replace(/\s+/g, ' ').trim()).filter(Boolean);
  return cleaned.slice(0, 3);
}

export function formatRecallBlock(items: any, precision = DEFAULT_RECALL_SCORE_PRECISION, sessionId?: string, queryId?: string) {
  if (!Array.isArray(items) || items.length === 0) return '';
  const attrs = [sessionId && `session_id="${sessionId}"`, queryId && `query_id="${queryId}"`].filter(Boolean).join(' ');
  const lines = [attrs ? `<recall ${attrs}>` : '<recall>'];
  for (const item of items) {
    const score = Number.isFinite(item?.score_display) ? Number(item.score_display).toFixed(precision) : String(item?.score ?? '');
    const cues = readCueList(item);
    const cueText = cues.join(' · ').trim();
    lines.push(`${score} | ${item?.uri || ''}${cueText ? ` | ${cueText}` : ''}`);
  }
  lines.push('</recall>');
  return lines.join('\n');
}

export function normalizeSearchResults(data: any) {
  if (Array.isArray(data)) return data;
  if (Array.isArray(data?.results)) return data.results;
  return [];
}

export function normalizeKeywordList(values: any) {
  if (!Array.isArray(values)) return [];
  const seen = new Set();
  const out: string[] = [];
  for (const value of values) {
    const keyword = String(value || '').trim();
    if (!keyword) continue;
    const key = keyword.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(keyword);
  }
  return out;
}

export function normalizeUriList(items: any) {
  if (!Array.isArray(items)) return [];
  return [...new Set(items.map((item: any) => String(item?.uri || item || '').trim()).filter(Boolean))];
}
